### PR TITLE
Added custom json configs and improved autotiles

### PR DESCRIPTION
**Pull request by Vegita2**
This pull request adds two things
1. Custom JSON Configurations
2. Improved Autotiles

### Custom JSON Configurations

Users can now provide custom JSON configs for the map editor through mods.

#### Implementation Details:
- Configs from all mods are loaded and combined
- Mod configs override default configs

#### Usage:
1. Create a `map-editor` folder in your mods directory:
   ```
   CrossCode/assets/mods/[mod_name]/map-editor/
   ```
2. Add JSON files to override base map editor configs
   - Reference files: [Default Configs](https://github.com/CCDirectLink/crosscode-map-editor/tree/master/webapp/src/assets)

### Improved Autotiles

This update enhances autotile functionality and introduces a new autotile type.

#### Changes:
- Autotile definitions changed from strings to size objects
- New 4x4 autotile type implemented

#### Autotile Types:
1. Height 2 autotiles:
   - Width options: 8, 10, 12, 14
   - Example:
     

2. 4x4 autotiles:
   - Example:
     4x4 Autotile

I made a simple tool to help creating autotile definitions: https://github.com/Vegita2/cc-autotiler

### CODE REVIEW

File: webapp/src/app/components/widgets/event-widget/event-editor/event-helper.service.ts [hunk 1]
```diff
@@ -13,7 +14,8 @@ export class EventHelperService {
 	
 	constructor(
 		private eventRegistry: EventRegistryService,
-		private domSanitizer: DomSanitizer
+		private domSanitizer: DomSanitizer,
+		private jsonLoader: JsonLoaderService
```
Comment (2767mr) [inline webapp/src/app/components/widgets/event-widget/event-editor/event-helper.service.ts:18]: This seems to be unused

File: webapp/src/app/models/map-styles.ts [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
-};
+}
 
 export type MapStyle = {
 	sheet: string;
 	hasDoorMat?: boolean;
+	walls?: WallColors;
 } & {
 	[key: string]: any;
 };
```
Comment (2767mr) [inline webapp/src/app/models/map-styles.ts:11]: You might as well do this while you're at it
```suggestion
export interface MapStyle {
	sheet: string;
	hasDoorMat?: boolean;
	walls?: WallColors;
	[key: string]: any;
}
```

**Review by 2767mr — CHANGES_REQUESTED** [review]
It works exactly as I would expect. I think this will actually have a major impact on mod developers.

Just some minor nitpicks (mostly internal stuff).

One small bug I found:
After selecting a different layer, the selection box is not rendered properly.

We should probably document which .json file does what and how it is structured but I wouldn't block the PR on that.

**Comment by Vegita2** [comment]
> One small bug I found:
After selecting a different layer, the selection box is not rendered properly.

Can you elaborate a bit? I don't know what you mean

**Comment by 2767mr** [comment]
> > One small bug I found:
> > After selecting a different layer, the selection box is not rendered properly.
> 
> Can you elaborate a bit? I don't know what you mean

Start: 
image

After selecting "Upper":
image

File: common/src/controllers/api.ts [hunk 3]
```diff
@@ -1,14 +1,21 @@
 import { fsPromise, pathPromise } from '../require.js';
 import { saveFile as save } from './saveFile.js';
 
+
+export interface ConfigExtension<T> {
```
[the remark this answers is not in the document]
Comment (Vegita2) [reply common/src/controllers/api.ts:5]: Asked claude ai to suggest some names.

- ModConfigExtension
- ModEditorConfig
- ExternalModDefinition
- ModEditorExtension
- CustomModTypeDefinition
- ModTypeAddition

Do you have any preference?

[the remark this answers is not in the document]
Comment (2767mr) [reply common/src/controllers/api.ts:5]: ModEditorConfig would work for me

File: webapp/src/app/components/dialogs/map-settings/map-settings.component.ts [hunk 4]
```diff
@@ -63,6 +63,8 @@ export class MapSettingsComponent {
 			y: settings.mapHeight
 		});
 		
+		this.events.updateEntities.next(true);
```
[the remark this answers is not in the document]
Comment (Vegita2) [reply webapp/src/app/components/dialogs/map-settings/map-settings.component.ts:66]: It fixes `Destructible` entities. Their rendering depends on the `mapStyle` property

File: webapp/src/app/services/phaser/BaseTileDrawer.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+	
+	private onMouseRightUp() {
+		this.rightPointerDown = false;
+		if (!this.layer) {
+			return;
+		}
+		
+		// cancel current selection when out of bounds
+		const phaserLayer = this.layer.getPhaserLayer();
+		if (!this.rightClickStart || !this.rightClickEnd || !phaserLayer) {
```
[the remark this answers is not in the document]
Comment (Vegita2) [reply webapp/src/app/services/phaser/BaseTileDrawer.ts:183]: should be fixed, but would be good if you could try it out a bit. There could be some edge cases left

**Comment by Vegita2** [comment]
fixed issues

**Review by 2767mr — APPROVED** [review]
Tested it again and found 2 small nitpicks:
* The history texts are above the tile drawer
image
* The tile drawer does not resize itself anymore, so on 1080p monitors it does not fit all of the arid tileset at once

**Comment by Vegita2** [comment]
> The tile drawer does not resize itself anymore, so on 1080p monitors it does not fit all of the arid tileset at once
I don't know what you mean. The tile selector always had a fixed size

**Comment by 2767mr** [comment]
I was sure that it did resize at some point and did some digging. Apparently, 5 years ago in 0.4.0 it would resize itself to fit the tileset.

Obviously, I do not expect you to add it back in this PR (or at all for that matter)

**Pull request merged**